### PR TITLE
The Real Guice

### PR DESCRIPTION
**Pull request by s0cks**
### Description of the Change

I can't resist helping stop bad patterns from continuing, this PR is to add Guice to the app. Guice is a inversion of control based dependency injection framework. This will allow developers to construct objects more fluidly, and drastically de-couple the application

### Testing

Local regression testing on changed features. More is welcome

### Related Issues

Many

### CODE REVIEW

File: src/main/java/com/atlauncher/inject/AppModule.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package com.atlauncher.inject;
+
+import com.atlauncher.App;
+import com.atlauncher.data.Settings;
+import com.google.inject.AbstractModule;
+import com.google.inject.Provides;
+
+public final class AppModule extends AbstractModule{
```
Comment (Doomsdayrs) [inline src/main/java/com/atlauncher/inject/AppModule.java:25]: Documentation?

File: src/main/java/com/atlauncher/inject/ArgumentsModule.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+package com.atlauncher.inject;
+
+import com.google.inject.AbstractModule;
+import com.google.inject.Provides;
+import joptsimple.OptionParser;
+import joptsimple.OptionSet;
+
+import javax.inject.Named;
+
+public final class ArgumentsModule extends AbstractModule{
```
Comment (Doomsdayrs) [inline src/main/java/com/atlauncher/inject/ArgumentsModule.java:27]: Documentation?

File: src/main/java/com/atlauncher/inject/ArgumentsParserProvider.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package com.atlauncher.inject;
+
+import joptsimple.OptionParser;
+
+import javax.inject.Provider;
+import java.util.Arrays;
+
+public final class ArgumentsParserProvider implements Provider<OptionParser> {
```
Comment (Doomsdayrs) [inline src/main/java/com/atlauncher/inject/ArgumentsParserProvider.java:25]: Documentation?

File: src/main/java/com/atlauncher/inject/ArgumentsProvider.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+package com.atlauncher.inject;
+
+import joptsimple.OptionParser;
+import joptsimple.OptionSet;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Provider;
+
+public final class ArgumentsProvider implements Provider<OptionSet>{
```
Comment (Doomsdayrs) [inline src/main/java/com/atlauncher/inject/ArgumentsProvider.java:27]: Documentation?

**Review by Doomsdayrs — CHANGES_REQUESTED** [review]
The PR looks great, and showcases how DI can be beneficial, in a minor way.

Regardless, with a bit of documentation it should be ready for review by Ryan

**Comment by Doomsdayrs** [comment]
Since this is adding a new dependency, what is the cost on final jar size?

File: build.gradle [hunk 5]
```diff
@@ -43,6 +43,7 @@ repositories {
 }
 
 dependencies {
+    implementation 'com.google.inject:guice:5.1.0'
```
Comment (Doomsdayrs) [inline build.gradle:46]: Since this is adding a dependency, what is the jar size cost?

Comment (s0cks) [reply build.gradle:46]: > jar (759 KB)
https://mvnrepository.com/artifact/com.google.inject/guice/5.1.0


I mean jar size shouldn't matter. If it does we should fix that.
~ Tazz

Comment (Doomsdayrs) [reply build.gradle:46]: Jar size is important, it is the size of the application after all.

Furthermore I am not talking about the dependency jar, I am considering the dependencies final impact to the launcher jar. 

For example, dbus lead to an approximately 5mb increase in file size due to its implied dependencies.

**Pull request closed without merging**